packages/dapp-svelte-wallet/api/src/lib-wallet.js

```diff
@@ -1250,47 +1245,129 @@ export function makeWallet({
+    getAll: () => installationMapping.petnameToVal.entries(),
+    add: (petname, thing) => addInstallation(petname, thing),
+  });
+
+  const instanceManager = harden({
+    rename: (petname, instance) => {
+      instanceMapping.renamePetname(petname, instance);
+      // We don't wait for the update before returning.
+      updateAllState();
+    },
+    get: petname => instanceMapping.petnameToVal.get(petname),
+    getAll: () => instanceMapping.petnameToVal.entries(),
+    add: (petname, instance) => {
+      // We currently just add the petname mapped to the instance
+      // value, but we could have a list of known instances for
+      // possible display in the wallet.
+      petname = instanceMapping.suggestPetname(petname, instance);
+      // We don't wait for the update before returning.
+      updateAllState();
+    },
+  });
+
+  const issuerManager = harden({
+    rename: (petname, issuer) => {
+      assert(
+        brandTable.hasByIssuer(issuer),
+        `issuer has not been previously added`,
+      );
+      const brandRecord = brandTable.getByIssuer(issuer);
+      brandMapping.renamePetname(petname, brandRecord.brand);
+      updateAllState();
+    },
+    get: petname => {
+      const brand = brandMapping.petnameToVal.get(petname);
+      return brandTable.getByBrand(brand).issuer;
+    },
+    getAll: () => {
+      // TODO: what is the return value?
+      return brandMapping.petnameToVal.entries().map(([petname, brand]) => {
+        const { issuer } = brandTable.getByBrand(brand);
+        return [petname, issuer];
+      });
+    },
+    add: async (petname, issuerP) => {
+      const { brand, issuer } = await brandTable.initIssuer(issuerP);
+      if (!issuerToBoardId.has(issuer)) {
+        const issuerBoardId = await E(board).getId(issuer);
+        issuerToBoardId.init(issuer, issuerBoardId);
+      }
+      brandMapping.suggestPetname(petname, brand);
+      updateAllIssuersState();
+      return issuer;
+    },
+  });
+
+  function getInstallationManager() {
+    return installationManager;
   }
 
-  async function renameInstance(petname, instance) {
-    instanceMapping.renamePetname(petname, instance);
-    await updateAllState();
-    return `instance ${q(petname)} successfully renamed in wallet`;
+  function getInstanceManager() {
+    return instanceManager;
   }
 
-  async function renameInstallation(petname, installation) {
-    installationMapping.renamePetname(petname, installation);
-    await updateAllState();
-    return `installation ${q(petname)} successfully renamed in wallet`;
+  function getIssuerManager() {
+    return issuerManager;
   }
 
-  function getIssuer(petname) {
-    const brand = brandMapping.petnameToVal.get(petname);
-    return brandTable.getByBrand(brand).issuer;
+  async function findAllInvitationAmount(invitationDetailsCriteria) {
+    const zoeInvitationPurse = getPurse(ZOE_INVITE_PURSE_PETNAME);
+    const invitationAmount = await E(zoeInvitationPurse).getCurrentAmount();
```

Comment:
This creates a painful race condition for its callers that I previously had to work very carefully around: what if the invitation has not yet arrived in the invitation purse?  In unit tests, this almost never happens, but in a real network, it happens almost every time.

To me, it would be better to implement this functionality mostly in the helper library, by giving it a more flexible tool in the wallet that it could use, like:

```js
/**
 * @typedef {[Petname, Amount]} PurseAndBalance A pair of a purse petname and an amount
 */
/**
 * @callback GetBrandBalanceNotifier
 * @param {Petname} brandPetname the petname for the brand of the purses to match
 * @returns {Notifier<Array<PurseAndBalance>>} a notifier for a list of purse petnames matching that brand and their recent balances
 */

Then build the matching logic into that helper function, which basically ends up being a notifier transform from `GetBrandBalanceNotifier`.

Such a primitive would simplify several dapp UIs, not just invites.

Reply:
There is definitely a race condition here, but it's not a new one. I think we're both very familiar with this one! See https://github.com/Agoric/agoric-sdk/blob/1587c0435ede413a3e9fdefbb24564d8e99fe27d/packages/dapp-svelte-wallet/api/src/lib-wallet.js#L694 which does the same thing, and which I wrote a hacky work-around for by requiring dapps to await the deposit of the invitation before sending the `addOffer` message. This function adds no new race conditions, as far as I know.

Long term, I'd like to tackle the race condition with an `eventualWithdraw` function added to the purse in `issuer.js`. This would build on [this PR](https://github.com/Agoric/agoric-sdk/pull/2157), which has some problems that I think would be solved if we know the code is in the same vat as the purse.

For the meantime though, I will move this code into the deploy-script-support package.